CaptainAgent PR Part 1: Adding blog post and document (#27)
ag2ai/ag2: website/blog/2024-11-15-CaptainAgent/index.mdx

---
title: "Introducing CaptainAgent for Adaptive Team Building"
authors:
  - LinxinS97
  - jialeliu
  - jieyuz2
tags: [LLM, GPT, AutoBuild]
---
![Illustration of how CaptainAgent build a team](img/overall.png)

**TL;DR**
- We introduce CaptainAgent, an agent equipped with the capability to adaptively assemble a team of agents through retrieval-selection-generation process to handle complex tasks via the [`nested chat`](https://ag2ai.github.io/ag2/docs/tutorial/conversation-patterns#nested-chats) conversation pattern in AG2.
- CaptainAgent supports all types of `ConversableAgents` implemented in Autogen.

# Introduction

> Assembling a team of LLM agents is easy, but dynamically choosing the right agents into a group chat according to the task is hard. In this blog, we introduce **CaptainAgent**, an agent that can automatically build a team of agents to fulfill diverse, complex task requirements.

Reviewer: ```suggestion
Given an ad-hoc task, dynamically assembling a group of agents capable of effectively solving the problem is a complex challenge. In many cases, we manually design and select the agents involved.  In this blog, we introduce **CaptainAgent**, an intelligent agent that can autonomously assemble a team of agents tailored to meet diverse and complex task requirements.
```

> We design a two-step workflow for CaptainAgent to build a high quality team and solve the problem:

Reviewer: ```suggestion
CaptainAgent iterates over the following two steps until the problem is successfully solved.
```

- (**Step 1**) CaptainAgent will first break down the task, list several roles needed for each subtask, and then create a team of agents accordingly by retrieval, selection, and generation. Each of these will be equipped with predefined tools retrieved from the tool library.
![Building workflow](img/build.png)

> - (**Step 2**) this team of agents will attempt to solve the subtask via conversation with the free-form tool using. Once it's done, a reflector LLM will provide CaptainAgent with a reflection report for it to decide whether to adjust the team or subtask instruction or to terminate and output the results.

Reviewer: ```suggestion
- (**Step 2**) For each subtask, the corresponding team of agents will jointly solve it. Once it's done, a summarization and reflection step will be triggered to generate a report based on the multi-agent conversation history.  Based on the report, CaptainAgent will decide whether to adjust the subtasks and corresponding team or to terminate and output the results.
```

> - (**Step 2**) For each subtask, the corresponding team of agents will jointly solve it. Once it's done, a summarization and reflection step will be triggered to generate a report based on the multi-agent conversation history.  Based on the report, CaptainAgent will decide whether to adjust the subtasks and corresponding team or to terminate and output the results.

Reviewer: ```suggestion
- (**Step 2**) For each subtask, the corresponding team of agents will jointly solve it. Once it's done, a summarization and reflection step will be triggered to generate a report based on the multi-agent conversation history.  Based on the report, CaptainAgent will decide whether to adjust the subtasks and corresponding team (go to Step 1) or to terminate and output the results.
```

![Building workflow](img/chat.png)

> We also provide an agent library and a tool library for CaptainAgent to choose from when building the team. In the following section, we demonstrate how to use CaptainAgent with or without the provided library.

Reviewer: ```suggestion
The design of CaptainAgent allows it to leverage agents and tools from a pre-specified agent library and tool library. In the following section, we demonstrate how to use CaptainAgent with or without the provided library.
```

> # Using CaptainAgent without libraries

Reviewer: ```suggestion
# Using CaptainAgent without pre-specified agent/tool libraries
```

> CaptainAgent can serve as a drop-in replacement for the general AssistantAgent. To use CaptainAgent, we need to instantiate a CaptainAgent and a UserProxy, then start a chat. To do that we just need to configure to the settings of group chat instantiated by CaptainAgent.

Reviewer: ```suggestion
CaptainAgent can serve as a drop-in replacement for the general `AssistantAgent` class in AG2. To do that we just need to add a few lines of configurations for the group chat involved.
```

> Without the agent library and tool library, CaptainAgent will automatically generate a set of agents into a group chat.

Reviewer: ```suggestion
Without the agent library and tool library, CaptainAgent will automatically generate a set of agents into a group chat.
```
```suggestion
When an agent library is not provided, CaptainAgent will automatically generate a set of agents from scratch to form a group chat.
```

```python
from autogen.agentchat.contrib.captain_agent import CaptainAgent

general_llm_config = {
    "temperature": 0,
    "config_list": autogen.config_list_from_json("OAI_CONFIG_LIST", filter_dict={"model": ["gpt-4-1106-preview"]}),
}

nested_mode_config = {
    "autobuild_init_config": {
        "config_file_or_env": "OAI_CONFIG_LIST",
        "builder_model": "gpt-4-1106-preview",
        "agent_model": "gpt-4-1106-preview",
    },
    # this is used to configure the autobuild building process
    "autobuild_build_config": {
        "default_llm_config": {"temperature": 1, "top_p": 0.95},
        "code_execution_config": {"timeout": 300, "work_dir": "groupchat", "last_n_messages": 1},
        "coding": True,
    },
    "group_chat_config": {"max_round": 15},
    "group_chat_llm_config": general_llm_config.copy(),
    "max_turns": 3,
}

## build agents
captain_agent = CaptainAgent(
    name="captain_agent",
    llm_config=general_llm_config,
    nested_mode_config=nested_mode_config,
)

> user_proxy = CaptainUserProxyAgent(

Reviewer: ```suggestion
user_proxy = UserProxyAgent(
```

    name="user_proxy",
    code_execution_config={"use_docker": False},
)
query = "Let's play game of 24. Given 4 numbers, you need to use +, -, *, / to get 24. The numbers are 2, 2, 7, 12."
result = user_proxy.initiate_chat(captain_agent, message=query)
```

> # Using CaptainAgent with libraries

Reviewer: ```suggestion
# Using CaptainAgent with pre-specified agent/tool libraries
```

> To use CaptainAgent with libraries, we just need to specify the path to the tool library and agent library. The two libraries are independent, so you can choose to use one of the libraries or both.

Reviewer: ```suggestion
To use CaptainAgent with pre-specified agent/tool libraries, we just need to specify the path to the agent library and tool library. The two libraries are independent, so you can choose to use one of the libraries or both.
```

> The tool library we provide requires subscribing to specific apis, please refer to the [docs](https://ag2ai.github.io/ag2/docs/topics/captainagent) for details. The following example does not necessarily requires tool usage.

Reviewer: ```suggestion
The tool library we provide requires subscribing to specific APIs, please refer to the [docs](https://ag2ai.github.io/ag2/docs/topics/captainagent) for details. The following example does not necessarily require tool usage, so it's fine if you are subscribing to them. 

To use agents from an agent library, you just need to specify a `library_path` sub-field or a `autobuild_tool_config` field in CaptainAgent's configuration. 
```

general_llm_config = {
    "temperature": 0,
    "config_list": autogen.config_list_from_json("OAI_CONFIG_LIST", filter_dict={"model": ["gpt-4-1106-preview"]}),
}

nested_mode_config = {
    "autobuild_init_config": {
        "config_file_or_env": "OAI_CONFIG_LIST",
        "builder_model": "gpt-4-1106-preview",
        "agent_model": "gpt-4-1106-preview",
    },
    # this is used to configure the autobuild building process
    "autobuild_build_config": {
        "default_llm_config": {"temperature": 1, "top_p": 0.95},
        "code_execution_config": {"timeout": 300, "work_dir": "groupchat", "last_n_messages": 1},
        "coding": True,
        "library_path": "captainagent_expert_library.json"
    },
    "autobuild_tool_config": {
        "tool_root": "default",  # this will use the tool library we provide
        "retriever": "all-mpnet-base-v2",
    },
    "group_chat_config": {"max_round": 10},
    "group_chat_llm_config": general_llm_config.copy(),
    "max_turns": 3
}

## build agents
captain_agent = CaptainAgent(
    name="captain_agent",
    llm_config=general_llm_config,
    nested_mode_config=nested_mode_config,
)

> user_proxy = CaptainUserProxyAgent(

Reviewer: ```suggestion
user_proxy = UserProxyAgent(
```

---
title: "Introducing Captain Agent for Adaptive Team Building"
authors:
  - LinxinS97
  - jialeliu
  - jieyuz2
tags: [LLM, GPT, AutoBuild]
---
![Illustration of how Captain Agent build a team](img/overall.png)

**TL;DR**

> - We introduce Captain Agent, an agent equipped with the capability to adaptively assemble a team of agents through retrieval-selection-generation process to handle complex tasks via nested chat.

Reviewer: ```suggestion
- We introduce Captain Agent, an agent equipped with the capability to adaptively assemble a team of agents through retrieval-selection-generation process to handle complex tasks via the [`nested chat`](https://ag2ai.github.io/ag2/docs/tutorial/conversation-patterns#nested-chats) conversation pattern in AG2.
```

> - We introduce Captain Agent, an agent equipped with the capability to adaptively assemble a team of agents through retrieval-selection-generation process to handle complex tasks via nested chat.

Reviewer: Captain Agent or CaptainAgent?

Author: I have fixed the term to CaptainAgent

We design a two-step workflow for Captain Agent to build a high quality team and solve the problem:
- (**Step 1**) Captain Agent will first identify a subtask instructed by our prompt, list several roles needed for this subtask, and then create a team of agents accordingly by retrieval, selection, and generation. Each of these will be equipped with predefined tools retrieved from the tool library.
![Building workflow](img/build.png)
- (**Step 2**) this team of agents will attempt to solve the subtask via conversation with the free-form tool using. Once it's done, a reflector LLM will provide Captain Agent with a reflection report for it to decide whether to adjust the team or subtask instruction or to terminate and output the results.
![Building workflow](img/chat.png)

We provide an agent library and a tool library for Captain Agent to choose from when building the team. In the following section, we demonstrate how to use CaptainAgent.

> # Using Captain Agent w/o libraries

Reviewer: The previous sentence mentioned "provide an agent library and a tool library for Captain Agent to choose from"
however, we are talking about using Captain Agent without libraries.

If my understanding is correct, we should mention that there are two ways to use CaptainAgent.
